pkg/jx/cmd/delete_aws.go

```diff
@@ -95,13 +97,27 @@ func (o *DeleteAwsOptions) Run() error {
 	}
 	for _, internetGateway := range internetGateways.InternetGateways {
 		if len(internetGateway.Attachments) > 0 {
-			_, err = svc.DetachInternetGateway(&ec2.DetachInternetGatewayInput{InternetGatewayId: internetGateway.InternetGatewayId, VpcId: aws.String(vpcid)})
-			if err != nil {
-				return err
+			detachAttemptsLeft := 11
```

Comment:
why `11` :)?

Reply:
I would recommend to use this retry mechanism https://github.com/jenkins-x/jx/blob/0ccc1304ed636279ea5b3860ee6a271c8ca79315/pkg/jx/cmd/common.go#L540

Reply:
Good point. I've converted "magic number" to constant :) . And it is 10 attempts, because it is reasonable number from my tests against dead VPC instances :) .

Reply:
When it comes to `retry()` it doesn't really handle my case. The thing is that if AWS return any error other than unmapping error, I'd like command to fail immediately. I want to retry only for certain type of error and fail immediately for all the others.

If we could consider extending `retry(..., call func() error) (err error)` signature to be more like `retry(..., call func() (err error, fatalError error)) (err error, fatalError error)` then it will cover my use case, because it would be possible to distinguish between recoverable callback error which should be retried and fatal error which should make retry to stop immediately and return error.

I propose to keep my retry logic as it is right now and then maybe create another PR in which we could extend `retury` to support fatal errors?

Reply:
I got your point. You can define a new retry function like this:

```Go
type FataError struct {
     E error
}

func (err *FataError) Error() string {
     return fmt.Sprintf("fatal error: %s", err.E.Error())
}

func (o *CommonOptions) retryUntilFatalError(attempts int, sleep time.Duration, call func() error) (err error) {
	for i := 0; ; i++ {
		err = call()
		if err == nil {
			return
		}
                
                ok, err := err.(FatalError)
                if ok {
                      return err
                }

		if i >= (attempts - 1) {
			break
		}

		time.Sleep(sleep)

		log.Infof("retrying after error:%s\n", err)
	}
	return fmt.Errorf("after %d attempts, last error: %s", attempts, err)
}
```

And then re-use it for you case. 

I am pretty sure that we will have this scenario in future.

Reply:
I'm just wondering, because we have only 12 places in jx codebase where `retry` is used, maybe we could just refactor existing function? That new version of function can cover all the existing use cases as it is superset of the old retry function.

Maybe a plan to go would be to create `retryUntilFatalError` function now and use it in my case, but also mark `retry` as deprecated? And migrate old codecase to use `retryUntilFatalError` instead of `retry` in next PRs?

Reply:
sounds good!

Reply:
Pushed changes. Can you take a look at those now? Thanks!